### PR TITLE
Printing state

### LINKED ISSUE #548: Print stateful components on interpreter output
We should be outputting the internal states of stateful components such as the memories and registers, particularly the former as they have state that cannot be fully observed through output ports. This could look like a JSON dump similar to the ports, though it would be nice to have something a little prettier to look at.

### PR DESCRIPTION
**Pull request by EclecticGriffin**
A cute little update to the serializer to output the state contained in stateful cells (i.e. memories & registers). It only does this when such cells are present, though generally there's at least one register.

We could remove registers from the output if we feel they're sufficiently represented in the port state, but I think it's nice to have them there.

Closes #548

### CODE REVIEW

File: interp/tests/simple_tests/simple_neq.expect [hunk 1]
```diff
@@ -1,25 +1,38 @@
...
-      "right": 10
-    },
-    "reg0": {
-      "clk": 0,
-      "done": 1,
-      "in": 0,
-      "out": 0,
-      "write_en": 1
+[
+  [
```
Comment (rachitnigam) [inline interp/tests/simple_tests/simple_neq.expect:2]: I would've expected this to be an object with "Ports:" and "Memory State:" being field names. I would also align the fields names to be similar to `fud --to dat` (so naming the field `memories` and the other field name that I don't remember).

Comment (EclecticGriffin) [reply interp/tests/simple_tests/simple_neq.expect:2]: Yeah a dummy struct probably makes more sense, I'll change that in a bit

Comment (EclecticGriffin) [reply interp/tests/simple_tests/simple_neq.expect]: Done: see most recent commit

**Comment by rachitnigam** [comment]
LGTM! As we try to bring the input and output formats closer to those of fud, it might make sense to stratify the representation even more and have a registers field separate from memories. We can do it when we start running the correctness tests with the interpreter.

**Review by rachitnigam — APPROVED** [review]

**Review by sampsyo — APPROVED** [review]
Seems great; I think this is worth merging right away! A couple of thoughts for the future:

- I wonder whether state serialization (like execution) might eventually want to be "decentralized" so it's, like, a trait function for each primitive. That could avoid the big `match` on primitive kinds (replacing it with a `dyn` invocation) and make it easier to add new primitives.
- I notice that memories seem to get serialized as strings, e.g., `"5"` for scalars and `"[4]"` for arrays. Perhaps it would make more sense someday to serialize them as JSON numbers/arrays, e.g., `5` and `[4]`?
- A tricky thing we will need to deal with eventually is that we (intentionally) don't know the data types being stored in the memories. That makes it impossible to be 100% compatible with fud-style inputs and outputs because fud is responsible for converting numerical formats (as it should be). So we will probably want to someday rely on fud again to convert between the basic "always-`u64`" format with custom data types. In other words, we could have the interpreter dump "raw" data and rely on fud to covert this JSON output to the same JSON output it produces from Verilator simulation. This frees us from needing to replicate that output format exactly using the interpreter alone.

**Pull request merged**